testssl.sh

```diff
@@ -8869,13 +8992,71 @@ sslv2_sockets() {
      return $ret
 }
 
+# arg1: supported groups extension
+# arg2: "all" - process full response (including Certificate and certificate_status handshake messages)
+#       "ephemeralkey" - extract the server's ephemeral key (if any)
+# Given the supported groups extension, create a key_share extension that includes a key share for
+# each group listed in the supported groups extension.
+generate_key_share_extension() {
+     local supported_groups
+     local -i i len supported_groups_len group
+     local extn_len list_len
+     local key_share key_shares=""
+
+     supported_groups="${1//\\x/}"
+     [[ "${supported_groups:0:4}" != "000a" ]] && return 1
+
+     supported_groups_len=${#supported_groups}
+     [[ $supported_groups_len -lt 16 ]] && return 1
+
+     len=2*$(hex2dec "${supported_groups:4:4}")
+     [[ $len+8 -ne $supported_groups_len ]] && return 1
+
+     len=2*$(hex2dec "${supported_groups:8:4}")
+     [[ $len+12 -ne $supported_groups_len ]] && return 1
+
+     for (( i=12; i<supported_groups_len; i=i+4 )); do
+          group=$(hex2dec "${supported_groups:i:4}")
+          # If the Supported groups extensions lists more than one group,
+          # then don't include the larger key shares in the extension.
+          [[ $i -gt 12 ]] && [[ $group -gt 256 ]] && continue
+
+          # Versions of OpenSSL prior to 1.1.0 cannot perform operations
+          # with X25519 keys, so don't include the X25519 key share
+          # if the server's response needs to be decrypted and an
+          # older version of OpenSSL is being used.
+          [[ $i -gt 12 ]] && [[ $group -eq 29 ]] && [[ "$2" == "all" ]] && \
+               [[ $OSSL_VER_MAJOR.$OSSL_VER_MINOR != "1.1.0"* ]] && \
+               [[ $OSSL_VER_MAJOR.$OSSL_VER_MINOR != "1.1.1"* ]] && \
```

Comment:
According to the ChangeLog, LibreSSL 2.5.1 added "Curve25519 Key Exchange support." However, I tried both 2.5.3 and 2.6.2 and couldn't get anything to work with X25519. So, unless someone has more success with LibreSSL than I did, I think this can stay as only OpenSSL 1.1.0 and 1.1.1.